### PR TITLE
New component: sumologicprocessor

### PR DESCRIPTION
**Pull request by sumo-drosiek**
**Description:**

Adds Sumo Logic Processor

The Sumo Logic processor (config name: `sumologic`)
modifies the metadata on logs, metrics and traces sent to [Sumo Logic][sumologic_webpage]
so that the Sumo Logic [apps][sumologic_apps] can make full use of the ingested data.

[sumologic_webpage]: https://www.sumologic.com
[sumologic_apps]: https://www.sumologic.com/applications/

**Link to tracking Issue:** #23946

**Testing:**

- Unit tests
- Component is being used by our customers

**Documentation:** `README.md`

### CODE REVIEW

**Comment by sumo-drosiek** [comment]
I may try to split that PR to smaller ones if it'll help. Now it is just moved from [Sumo Logic repository](https://github.com/SumoLogic/sumologic-otel-collector/tree/v0.80.0-sumo-0/pkg/processor/sumologicschemaprocessor)

**Comment by sumo-drosiek** [comment]
cc: @astencel-sumo

File: processor/sumologicprocessor/README.md
Comment (aboguszewski-sumo) [inline processor/sumologicprocessor/README.md:90]: For the record: these two features were created as a temporary solution for these issues: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14977 https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14946

**Comment by aboguszewski-sumo** [comment]
I don't know if it's needed, but I confirm my will to be a codeowner for this component.

File: processor/sumologicprocessor/metadata.yaml
Comment (djaglowski) [inline processor/sumologicprocessor/metadata.yaml:7]:
```suggestion
  distributions: [contrib, sumo]
```
I'm guessing this will be the case at first?

**Comment by sumo-drosiek** [comment]
@evan-bradley Could you take a look?

**Review by evan-bradley — COMMENTED** [review]
@sumo-drosiek This PR is fairly large, so I only did a surface pass. I will try to go back and do an in-depth review within a week.

File: processor/sumologicprocessor/README.md
Comment (sumo-drosiek) [reply processor/sumologicprocessor/README.md]: should be, but it seems I forget to do it :confused: 
Fixed now

**Comment by sumo-drosiek** [comment]
@evan-bradley 
> @sumo-drosiek This PR is fairly large, so I only did a surface pass

Thank you, I made requested changes

File: processor/sumologicprocessor/README.md
Comment (astencel-sumo) [inline processor/sumologicprocessor/README.md:1]:
```suggestion
# Sumo Logic Processor
```

We should stick to the "Sumo Logic processor" name

File: processor/sumologicprocessor/README.md
Comment (astencel-sumo) [inline processor/sumologicprocessor/README.md:13]:
```suggestion
The Sumo Logic processor (config name: `sumologic`)
```

File: .chloggen/drosiek-sumologic-schema-processor.yaml
Comment (astencel-sumo) [inline .chloggen/drosiek-sumologic-schema-processor.yaml:12]:
```suggestion
note: add Sumo Logic Processor
```

**Review by andrzej-stencel — APPROVED** [review]
This code has been used in production for many months as part of the Sumo Logic distro, as "Sumo Logic Schema" processor: https://github.com/SumoLogic/sumologic-otel-collector/tree/v0.83.0-sumo-0/pkg/processor/sumologicschemaprocessor. I believe it should be good to go, as long as contrib-specific requirements are met.

File: .github/dependabot.yml
Comment (sumo-drosiek) [inline .github/dependabot.yml:1139]:
```suggestion

```

**Comment by sumo-drosiek** [comment]
@evan-bradley Could you take a look? It is open quite long already :/

File: processor/sumologicprocessor/processor.go
Comment (evan-bradley) [inline processor/sumologicprocessor/processor.go:73]: I believe you could convert `procs` to a `[]zap.Bool` and then spread it here, something like: `logger.Info("", enabledProcessors...)`.

File: processor/sumologicprocessor/config.go
Comment (evan-bradley) [inline processor/sumologicprocessor/config.go:22]: Should the field name and config key name match more closely? "Patterns" and "prefixes" seem close but still distinct to me.

File: processor/sumologicprocessor/aggregate_attributes_processor.go
Comment (evan-bradley) [inline processor/sumologicprocessor/aggregate_attributes_processor.go:127]: Having an issue attached to this TODO would be nice.

**Review by evan-bradley — APPROVED** [review]
Overall this looks good to me, just a few nitpick comments.

**Comment by evan-bradley** [comment]
@sumo-drosiek Thanks for your patience. It looks like there are a few conflicts, if those are resolved we should be good to go.

File: processor/sumologicprocessor/aggregate_attributes_processor.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/aggregate_attributes_processor.go]: I added validation to config

File: processor/sumologicprocessor/config.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/config.go]: I unified them to prefixes

File: processor/sumologicprocessor/processor.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/processor.go]: Converted :)

File: processor/sumologicprocessor/aggregate_attributes_processor.go
Comment (evan-bradley) [reply processor/sumologicprocessor/aggregate_attributes_processor.go]: Just to confirm, does the config validation you added solve the issue mentioned in this comment? My understanding of the comment is that trimming prefixes from attribute keys can result in conflicting keys. For example `one_foo` and `two_foo` both have the same suffix `foo`, so trimming the prefixes from these keys can produce the same key. From what I can tell, the config validation only verifies that the user doesn't try to specify `pref` and `pref1` both as prefixes.

Comment (sumo-drosiek) [reply processor/sumologicprocessor/aggregate_attributes_processor.go:127]: Thats true. I created an issue in Sumo Logic repository, which I will move to this one, after the PR will be merged

**Comment by evan-bradley** [comment]
@sumo-drosiek Sorry for the churn, it looks like another file has changed in main. If you resolve the conflict, I'll get this merged.

**Comment by sumo-drosiek** [comment]
@evan-bradley no problem, quite a big PR :see_no_evil: 
Let's hope CI will be fine :)

**Comment by evan-bradley** [comment]
@sumo-drosiek Not sure what happened, but it looks like you dragged in a few extra commits with the latest merge from main. The diff still looks okay, so it may not matter after the changes are squashed, but if you can find an easy way to remove those it would be nice.

It looks like another change was merged in before I could take a look this morning. Unfortunately this is fairly common with new component PRs.  Could you please fix the conflicts?

**Comment by sumo-drosiek** [comment]
> Not sure what happened, but it looks like you dragged in a few extra commits with the latest merge from main.

Not sure as well, was just using `git merge origin/main` but something didn't happened as expected :/

Not sure if I can easily remove these comments. Not sure what'll happen

File: processor/sumologicprocessor/nesting_processor.go
Comment (MovieStoreGuy) [inline processor/sumologicprocessor/nesting_processor.go:248]: I suggest inverting the condition here to allow for early exiting.

File: processor/sumologicprocessor/translate_telegraf_metrics_processor.go
Comment (MovieStoreGuy) [inline processor/sumologicprocessor/translate_telegraf_metrics_processor.go:87]: Revert the condition to exist early.

**Review by MovieStoreGuy — COMMENTED** [review]
This PR is too large to effecitvely review, I suggest breaking this up into smaller sections as recommended by the contribution guidelines.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components

File: processor/sumologicprocessor/README.md
Comment (sumo-drosiek) [reply processor/sumologicprocessor/README.md:245]: https://github.com/SumoLogic/sumologic-otel-collector/issues/1263

Should we mention it in README?

File: processor/sumologicprocessor/README.md
Comment (sumo-drosiek) [reply processor/sumologicprocessor/README.md:86]: 1. Mark configuration as deprecated and add warning in case of using it and also instruction how to use new configuration
2. Remove support for the functionality, but leaving logs in order to avoid confusion with default missing configuration error
3. Remove logs and configuration option completely

File: processor/sumologicprocessor/README.md
Comment (sumo-drosiek) [reply processor/sumologicprocessor/README.md:125]: It is expected to rename the attribute, so combination of both

File: processor/sumologicprocessor/aggregate_attributes_processor.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/aggregate_attributes_processor.go:26]: Deduplication of what?

File: processor/sumologicprocessor/aggregate_attributes_processor.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/aggregate_attributes_processor.go:90]: Not at this moment

File: processor/sumologicprocessor/config.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/config.go:50]: > make this a immutable

How to do it?

This is declared as var to be easily accessible and give some clearness about default values, and because golang does not allow to create constant with `[]string{}` type

File: processor/sumologicprocessor/log_fields_conversion_processor.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/log_fields_conversion_processor.go:115]: > Also, should there be errors be returned in this function?

The should be an error in this implementation but we need to stick to interface

File: processor/sumologicprocessor/processor.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/processor.go:94]: Isn't going to be optimized by compiler? `subprocessor` is clearly used here to make code easier to read. I will change it to pointer if I'm wrong

Comment (sumo-drosiek) [reply processor/sumologicprocessor/processor.go:94]: I refactored it slightly to use range

Comment (aboguszewski-sumo) [reply processor/sumologicprocessor/processor.go]: @sumo-drosiek Does it work correctly with range? I don't remember exactly, but if this was written this way, then probably there was some issue behind using range.

Comment (sumo-drosiek) [reply processor/sumologicprocessor/processor.go]: Yes, I think we missed it during refactor: https://github.com/SumoLogic/sumologic-otel-collector/pull/675

File: processor/sumologicprocessor/README.md
Comment (alewandowski-sumo) [reply processor/sumologicprocessor/README.md:110]: 1. No idea
2. Additional data point: AWS Cloud Watch metrics use `aws/ec2` format for `namespace`

File: processor/sumologicprocessor/README.md
Comment (MovieStoreGuy) [reply processor/sumologicprocessor/README.md:271]: TIL

File: processor/sumologicprocessor/aggregate_attributes_processor.go
Comment (MovieStoreGuy) [reply processor/sumologicprocessor/aggregate_attributes_processor.go:26]: If a user defines the same attribute with different prefixes, is that fine?

File: processor/sumologicprocessor/config.go
Comment (MovieStoreGuy) [reply processor/sumologicprocessor/config.go:47]: > How to do it?

Define a read only interface. Another way to get around the race detector would be to define these per configuration created which still provides context of the vars but doesn't allow them to be shared across configurations.

**Review by MovieStoreGuy — APPROVED** [review]

File: processor/sumologicprocessor/README.md
Comment (evan-bradley) [reply processor/sumologicprocessor/README.md:271]: Just a minor note, it may be worth changing "a priori" to "beforehand" considering the global audience of the Collector who may not be familiar with common Latin phrases.

File: processor/sumologicprocessor/aggregate_attributes_processor.go
Comment (sumo-drosiek) [reply processor/sumologicprocessor/aggregate_attributes_processor.go:26]: Yes, we didn't observed any issue so far except the undefined behavior which we do not have decided what the proper solution should be. We are also waiting to replace this feature with Open Telemetry processors

**Comment by evan-bradley** [comment]
@sumo-drosiek Please rebase one more time and I'll merge this.

File: processor/sumologicprocessor/cloud_namespace_processor.go
Comment (evan-bradley) [inline processor/sumologicprocessor/cloud_namespace_processor.go:7]: I'm sorry, I didn't think to check for this yesterday: can this be easily replaced with `go.opentelemetry.io/collector/semconv`? Right now the processor uses `go.opentelemetry.io/collector/model` which doesn't receive releases anymore and has some old dependencies on the Go net and Google gRPC modules. There were just some CVEs in these modules and I want to make sure we keep the dependency tree clean.

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
